### PR TITLE
[10.x] Add a new `Numberable` helper

### PR DESCRIPTION
**Pull request by emmadesilva**
## Abstract

Building on https://github.com/laravel/framework/pull/48845, this class will work similarly to the `Stringable` class, but for numbers.

This was requested on the Discord and by several people on Twitter, and mentioned in https://github.com/laravel/framework/pull/48849

### Development

Feedback and contributions are much appreciated! I can be reached here, or in the `#internals` Discord.

- [x] Add accessors for the `Number` class
- [x] Add methods to interact with the value
- [x] Add `Number::of()` and `number()` shorthands
- [ ] See if we can find a better name for this helper? (Numerable has been suggested)

### CODE REVIEW

**Comment by rmunate** [comment]
https://github.com/laravel/framework/pull/48845#issuecomment-1817535149

**Comment by maartenpaauw** [comment]
Would be a great addition, but wouldn't it be better to make this class immutable like the stringable class?

**Comment by caendesilva** [comment]
> Would be a great addition, but wouldn't it be better to make this class immutable like the stringable class?

Thanks for the feedback! It definitely could be, but I went this route because I personally think modifying the values is more helpful for the kinds of use cases I see for this feature being used. But I am 100% open to changing it to be immutable if that turns out to be the consensus on what's best.

**Comment by bert-w** [comment]
This makes every equation like `($x + 5)/3` unreadable when you write it in functional fluent mode `Numberable($x)->add(5)->divide(3)`.

I would never use this, unless there's actually logic inside that would be worth abstractifying like vector or matrix math.

**Comment by caendesilva** [comment]
> This makes every equation like `($x + 5)/3` unreadable when you write it in functional fluent mode `Numberable($x)->add(5)->divide(3)`.
> 
> I would never use this, unless there's actually logic inside that would be worth abstractifying like vector or matrix math.

That's a great point, and a strong reason to use immutable versions. I'll make a patch tomorrow.

**Comment by Rizky92** [comment]
I'm not sure about this PR. `Stringable` works because there is a need for complex string manipulation fluently.

Your previous PR is great for converting unit of measurement or making numeric values readable, but not for manipulating it.

@bert-w's comments is a strong concern. Unless you provide another abstraction for working with arithmetics, I don't really see a good thing why this PR is necessary.

**Comment by caendesilva** [comment]
> I'm not sure about this PR. `Stringable` works because there is a need for complex string manipulation fluently.
> 
> Your previous PR is great for converting unit of measurement or making numeric values readable, but not for manipulating it.
> 
> @bert-w's comments is a strong concern. Unless you provide another abstraction for working with arithmetics, I don't really see a good thing why this PR is necessary.

Good points as well! Thanks for all the feedback! Going to try to refine this tomorrow, then open up for review to see what the team thinks. I appreciate your thoughts!

File: tests/Support/SupportNumberableTest.php [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        $this->assertSame($this->numberable(10)->percentage(), Number::percentage(10));
+        $this->assertSame($this->numberable(1.234)->percentage(2), Number::percentage(1.234, 2));
+        $this->assertSame($this->numberable(1.234)->percentage(2, 1), Number::percentage(1.234, 2, 1));
+        $this->assertSame($this->numberable(10000)->percentage(locale: 'de'), Number::percentage(10000, locale: 'de'));
+    }
+
+    public function testCurrency()
+    {
+        $this->assertSame($this->numberable(10)->currency(), Number::currency(10));
+        $this->assertSame($this->numberable(1.234)->currency(2), Number::currency(1.234, 2));
```
Comment (huangdijia) [inline tests/Support/SupportNumberableTest.php:148]: These two testcases should fail.

**Comment by gocanto** [comment]
Sincere and humbre question

Did you evaluate the possibility to use [Money/Money](https://github.com/moneyphp/money/tree/master)? - This lib is battle tested and offer this [functionality](https://github.com/moneyphp/money/blob/master/src/Number.php) out of the box.

cc @taylorotwell @driesvints

File: tests/Support/SupportNumberableTest.php (hunk 1 above)
Comment (caendesilva) [reply tests/Support/SupportNumberableTest.php:148]: My bad. Fixing it now.

**Comment by caendesilva** [comment]
> 🤝 Sincere and humble question,
> 
> Did you evaluate the possibility to use [Money/Money](https://github.com/moneyphp/money/tree/master)? - This lib is battle tested and offer this [functionality](https://github.com/moneyphp/money/blob/master/src/Number.php) out of the box.
> 
> cc @taylorotwell @driesvints

I haven't considered it for this feature, as I think these helpers are best used for quick formatting. If one needs something more robust, I think that's when you would reach for a library like yours. In my opinion, specialised code like that does not need to be in the Laravel core as we have packages to get the functionality when needed.

**Comment by caendesilva** [comment]
> I don't really think this is necessary yet. It's common to chain many string operations together when working with strings - I don't think the same is really the case with numbers.

Fair enough! It was requested by some people after my last PR, and it was fairly quick to make this one so I thought I'd throw it out there in case anyone finds it useful. Thanks for taking the time to review it!

**Pull request closed without merging**